### PR TITLE
Added Qthreads, MainWindow plotting for velocity with run/continue and pause user controls

### PR DESCRIPTION
**Pull request by rdriven**
README.txt contains more details about the contents of the branch.

### CODE REVIEW

File: datastore.py [hunk 1]
```diff
@@ -25,105 +27,150 @@ def __init__(self):
...
+                     extra={'sim_index': self._simulation_index})
+        temp = deepcopy(self._simulation_index)
+        self._lock.unlock()
+        return temp
 
     def get_walk_back_counter(self):
-        with self._lock:
-            return deepcopy(self._walk_back_counter)
+        self._lock.lockForRead()
+        temp =deepcopy(self._walk_back_counter)
```
Comment (mheidebr) [inline datastore.py]: add space

Comment (rdriven) [reply datastore.py]: Added space. :-)

File: simulation.py [hunk 2]
```diff
@@ -1,293 +1,524 @@
...
 
-    Returns:
-        Nothing (all data saved in the datastore)
+    def __init__(self, *args, **kwargs):
+        QWidget.__init__(self, parent = None)
+        print('Window: __init__')
+
+        self.data_store = DataStore()
+        print("MainWindow: Simulation Index = {}".format(self.data_store.get_simulation_index()))
+        #self.data_store.increment_simulation_index()
```
Comment (mheidebr) [inline simulation.py]: commented code

File: simulation.py [hunk 3]
```diff
@@ -1,293 +1,524 @@
...
 
-    Args:
-        data_store (DataStore): Thread safe storage for all simulation data
+class MainWindow(QWidget):
 
-    Returns:
-        Nothing (all data saved in the datastore)
+    def __init__(self, *args, **kwargs):
+        QWidget.__init__(self, parent = None)
+        print('Window: __init__')
```
Comment (mheidebr) [inline simulation.py:65]: Eventually I'd like all prints to become logging calls, we can define custom levels of logging and turn them on/off with arguments, for now #TODO this

File: simulation.py [hunk 4]
```diff
@@ -1,293 +1,524 @@
...
+
+        # create placeholder for the plots the SimulationThread will delivering 
+        # data into.
+        self.graphs = pg.GraphicsLayoutWidget(show=True, title="Race Sim plots")
+        self.graphs.resize(1000,540)
+        self.p1 = self.graphs.addPlot(name="Plot1", title="Velocity")        
+        self.p2 = self.graphs.addPlot(name="Plot1", title="Distance")        
+        self.p3 = self.graphs.addPlot(name="Plot1", title="Battery Power")        
+        self.p2.setXLink(self.p1)
+        #self.p2.setYLink(self.p1)
```
Comment (mheidebr) [inline simulation.py]: commented code

Comment (mheidebr) [reply simulation.py]: also 2 lines below

File: simulation.py [hunk 5]
```diff
@@ -1,293 +1,524 @@
...
+        self.simulationThread.simulationThreadSignal.connect(self.signalRcvFromSimulationThread)
+        self.plotRefreshThread.plotRefreshSignal.connect(self.signalPlotRefresh)
+
+        # TBD - what mechanism and what to do when SimulationThread or dies like
+        #       refresh GUI and save/close results file??
+        #self.simulationThread.finished.connect(self.simulationThreadFinished)
+        #self.simulationThread.terminated.connect(self.simulationThreadTerminated)
+
+
+        #Now that the SimulationThread has been created (but not yet running), connect the
```
Comment (mheidebr) [inline simulation.py]: space

File: simulation.py [hunk 6]
```diff
@@ -1,293 +1,524 @@
...
+        
+        # Create the instances of our worker threads
+        self.simulationThread = SimulationThread(self.data_store)
+        self.plotRefreshThread = PlotRefreshThread()
+
+        # Setup the SIGNALs to be received from the worker threads 
+        self.simulationThread.simulationThreadSignal.connect(self.signalRcvFromSimulationThread)
+        self.plotRefreshThread.plotRefreshSignal.connect(self.signalPlotRefresh)
+
+        # TBD - what mechanism and what to do when SimulationThread or dies like
```
Comment (mheidebr) [inline simulation.py]: TBD -> TODO

File: simulation.py [hunk 7]
```diff
@@ -1,293 +1,524 @@
...
+    @pyqtSlot()
+    def signalPlotRefresh(self):
+        index = self.data_store.get_simulation_index()
+        print("MainWindow:signalPlotRefresh Simulation Index = {}".format(index))
+        # create a new plot for every point simulated so far
+        x = [z+1 for z in range(self.data_store.get_simulation_index())]
+        _velocity = []
+        _distance = []
+        _battery_power = []
+        for z in x:
```
Comment (mheidebr) [inline simulation.py:164]: Add todo to move getting the velocity results from the beginning to the current index into the datastore, I think building the lists in the datastore will be faster (less locking/unlocking of thread lock)

File: simulation.py [hunk 8]
```diff
@@ -1,293 +1,524 @@
...
+
+        # Initialize the simulation universe
+        self._data_store = passed_data_store
+        self.initialize_race()
+
+        #print('SimulationThread: __init()__')
+        #print("SimulationThread: Simulation Index = {}".format(self._data_store.get_simulation_index()))
+
+        #connect some signals from the main window to us
+        #self.connect(self, QtCore.SIGNAL('To_End',self.processToEnd)
```
Comment (mheidebr) [inline simulation.py]: commented code

File: simulation.py [hunk 9]
```diff
@@ -1,293 +1,524 @@
...
+
+    def __del__(self):    
+        # Before a SimulationThread object is destroyed, we need to ensure that it stops processing. 
+        # For this reason, we implement the following method in a way that indicates to 
+        # the part of the object that performs the processing that it must stop, and waits 
+        # until it does so.
+        self.exiting = True
+        self.wait()
+
+        # rotational inertia estimation: http://www.hpwizard.com/rotational-inertia.html
```
Comment (mheidebr) [inline simulation.py]: move this comment to physics_equations.py

File: simulation.py [hunk 10]
```diff
@@ -1,293 +1,524 @@
...
+            #print('sim_index {} lap_results {}'.format(index, lap_results.motor_energy_list))
+
+            # TBD fix this 
+            #results.laps_per_pit_stop = car["battery_capacity"] / lap_results.motor_energy_list[-1]
+            results.lap_time = lap_results.end_velocity
+            results.lap_results = lap_results
+            self._data_store.set_race_results(results)
+
+
+    # TBD!!! remove data_store parameter and fix data_store references to use self._data_store
```
Comment (mheidebr) [inline simulation.py]: is this TBD done?

File: simulation.py [hunk 11]
```diff
@@ -1,293 +1,524 @@
...
+    def lap_velocity_simulation(self):
+        """Function calculates the velocity profile of a car with
+        car_properties on a track with track_properties. The car
+        starts with an ititial velocity of initial_velocity.
+
+        Args:
```
Comment (mheidebr) [inline simulation.py:285]: Remove

File: simulation.py [hunk 12]
```diff
@@ -1,293 +1,524 @@
...
+            data_store (DataStore): Thread safe storage for all simulation data
+
+        Returns:
+            Nothing (all data saved in the datastore)
+        """
+        # performance increases by assigning local functions
+        # https://towardsdatascience.com/10-techniques-to-speed-up-python-runtime-95e213e925dc
+        add_pys_result_to_datastore = self._data_store.add_physics_results_to_lap_results
```
Comment (mheidebr) [inline simulation.py]: change `add_pys_result_to_datastore` to `add_physics_result_to_datastore`

File: simulation.py [hunk 13]
```diff
@@ -1,293 +1,524 @@
...
+        # calculating should proceed in the simulation 
 
+        #print('SimulationThread: entering while() ')
+        while True:
+            # TBD We need to put the bulk of the work for Race Sim here
+            # with a state machine to run/pause/continue/step through calcs
+            print("SimulationThread:  running")
+            time.sleep(1)
+            if self.simulationComputing == True:
+                # The placeholder where the simulation should be doing the work
```
Comment (mheidebr) [inline simulation.py]: should this be removed?/what's the point of this whole commented block?

File: simulation.py [hunk 14]
```diff
@@ -1,293 +1,524 @@
...
+        self.exiting = False
 
-            current_lap_results = self._data_store.get_lap_results()
-            time = current_lap_results.time_profile
-            velocity = current_lap_results.velocity_profile
-            distance = current_lap_results.distance_profile
-            self._display_window.regraph(time, distance, velocity)
+        print('PlotRefreshThread: __init()__')
 
+        #connect some signals from the main window to us
```
Comment (mheidebr) [inline simulation.py]: commented code

File: simulation.py [hunk 15]
```diff
@@ -1,293 +1,524 @@
...
+
+    def run(self):        
+        # Note: This is never called directly. It is called by Qt once the
+        # thread environment with the thread's start() method has been setup, 
+        # and then runs "continuously" to do the work of the thread as it's main 
+        # processing loop 
+
+        print('PlotRefreshThread: entering while() ')
         while True:
+            #print("PlotRefreshThread: running")
```
Comment (mheidebr) [inline simulation.py]: commented code

File: simulation.py (hunk 9 above)
Comment (mheidebr) [reply simulation.py]: ended up moving this to electric car properties.py

File: simulation.py (hunk 10 above)
Comment (mheidebr) [reply simulation.py]: I think since the simulation runs this TBD is done

**Comment by mheidebr** [comment]
Before merging lets squash down the commits, probably 2, 1 for working changes, and 1 for linting

**Review by mheidebr — APPROVED** [review]

**Pull request merged**